### ISSUE TITLE
`needless_pass_by_ref_mut` false positive

### ISSUE THREAD

**Issue by aaron-ang** [rust-lang/rust-clippy#15920]
### Summary

Clippy is incorrectly detecting a needless mutable ref.

### Lint Name

needless_pass_by_ref_mut

### Reproducer

I tried this code:

```rust
#[cfg(feature = "python")]
fn foo(&mut self) -> Result<()> {
    let py_expr = PyExpr::from(self.udf_expr.inner().clone());
    let handle = Python::attach(|py| {
        Ok::<pyo3::Py<pyo3::PyAny>, PyErr>(
            py.import(pyo3::intern!(py, "hello.udf"))?
                .getattr(pyo3::intern!(py, "UdfHandle"))?
                .call1((py_expr,))?
                .unbind(),
        )
    })?;
    self.handle = Some(handle);
    Ok(())
}
```

I saw this happen:

```
warning: this argument is a mutable reference, but not used mutably
118 |     fn foo(&mut self) -> Result<()> {
    |                      ^^^^^^^^^ help: consider changing to: `&self`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_pass_by_ref_mut
    = note: `-D clippy::needless-pass-by-ref-mut` implied by `-D clippy::nursery`
    = help: to override `-D clippy::nursery` add `#[allow(clippy::needless_pass_by_ref_mut)]`
```

I expected to see this happen:
No errors because `self` needs to be mutable to assign its `handle` attribute.

### Version

```text
rustc 1.91.0-nightly (51ff89506 2025-09-02)
binary: rustc
commit-hash: 51ff895062ba60a7cba53f57af928c3fb7b0f2f4
commit-date: 2025-09-02
host: aarch64-apple-darwin
release: 1.91.0-nightly
LLVM version: 21.1.0
```
